internal/controller/cluster/reconciler.go

```diff
@@ -590,37 +608,58 @@ func (r *SingleClusterReconciler) isNewCluster() (bool, error) {
 	return len(statefulSetList.Items) == 0, nil
 }
 
-func (r *SingleClusterReconciler) hasClusterFailed() (bool, error) {
+// hasClusterFailed returns (failed, inGracePeriod, error)
+// failed: true if cluster has truly failed and needs recovery
+// inGracePeriod: true if cluster has failed pods but within grace period
+// error: any error during the check
+func (r *SingleClusterReconciler) hasClusterFailed() (failed, inGracePeriod bool, err error) {
 	isNew, err := r.isNewCluster()
 	if err != nil {
 		// Checking cluster status failed.
-		return false, err
+		return failed, inGracePeriod, err
 	}
 
 	if isNew {
 		// New clusters should not be considered failed.
-		return false, nil
+		return failed, inGracePeriod, nil
 	}
 
 	// Check if there are any pods running
 	pods, err := r.getClusterPodList()
 	if err != nil {
-		return false, err
+		return failed, inGracePeriod, err
 	}
 
 	for idx := range pods.Items {
 		pod := &pods.Items[idx]
-		if err := utils.CheckPodFailed(pod); err == nil {
+
+		podState := utils.CheckPodFailedWithGrace(pod, true)
+		if podState.State == utils.PodHealthy {
 			// There is at least one pod that has not yet failed.
 			// It's possible that the containers are stuck doing a long disk
 			// initialization.
 			// Don't consider this cluster as failed and needing recovery
 			// as long as there is at least one running pod.
-			return false, nil
+			return false, false, nil
 		}
+
+		// Pod is failed, check if it's in grace period
+		if podState.State == utils.PodFailedInGrace {
+			inGracePeriod = true
+		}
+	}
+
+	// If we reach here, all pods are failed
+	if inGracePeriod {
+		// Return grace period state
+		return failed, inGracePeriod, nil
+	}
+
+	if r.IsStatusEmpty() {
+		return true, inGracePeriod, nil
```

Comment:
inGracePeriod will always be false here, right?
If so, let’s return false explicitly for better clarity, similar to what we did on line 662.